modules/serverless-openshift-routes.adoc

```diff
@@ -12,9 +9,15 @@ If you want to configure a Knative service to use your TLS certificate on {produ
 
 .Procedure
```

Comment:
This procedure isn't working without doing the [Mapping a custom domain name to a Knative service](https://docs.openshift.com/container-platform/4.7/serverless/networking/serverless-domain-mapping.html) first. Without that, we get the 404 error from Kourier envoy (but TLS cert works well):

```
$ curl -iL http://heroku.com
HTTP/1.1 302 Found
Cache-Control: no-cache
Content-length: 0
Location: https://heroku.com/

HTTP/1.1 404 Not Found
date: Tue, 22 Jun 2021 00:28:44 GMT
server: envoy
content-length: 0
Set-Cookie: b64e0abd8f4e92f374f28865b191c274=86a7723b88e54f13e4a0e7f44a8c7934; path=/; HttpOnly; Secure; SameSite=None
Cache-control: private

```

We could include that page as a prerequisite, but then we would need to adjust this doc a little. When doing the [Mapping a custom domain name to a Knative service](https://docs.openshift.com/container-platform/4.7/serverless/networking/serverless-domain-mapping.html) it create an OpenShift Route (as results) named `route-<random-numbers>`. So, we need to instruct users to edit it instead, as creating additional Route will end up with that Route will not reconcile with error that given domain is already claimed:

```
route route-1900b946-ff1a-4373-8b93-b48663d28b30-373065613431 already exposes heroku.com and is older
```

But, that's also a problem as even if we edit that route (created foe DomainMapping), adding proper certificates, it gets rewrote by controller and certs are removed from it.

@nak3 could you respond to this issue?

Reply:
It seems that you are trying to create a custom domain by customizing OpenShift Route.
But [the doc](https://docs.openshift.com/container-platform/4.7/serverless/networking/serverless-domain-mapping.html) assumes only TLS customization on OCP Route.

If you want to configure custom domain + custom cert, we need DomainMapping (as you did) as well as adding the `serving.knative.openshift.io/disableRoute: true` label to DomainMapping.

Reply:
Okay, so you're saying this doc alone is to configure TLS certs. Fair, the title say so. But, how that's useful without configuring custom domain first?

Without custom domain you could only create certs for the same domain names as openshift is already providing, right?

And openshift is already providing certs for those names. User could just trust openshift CA or provide custom CA to create certs from (as we do on our test clusters).

If I'm right on this, then I'd say we should link those docs together, and instruct on domain mapping to skip creating the route with annotation you mentioned.

Reply:
> If I'm right on this, then I'd say we should link those docs together, and instruct on domain mapping to skip creating the route with annotation you mentioned.

Yes, I had a plan to add the instruction actually https://github.com/openshift/openshift-docs/pull/29621#discussion_r586059877
But there are some changes around TLS (e.g. we can now support custom Domain+TLS by net-istio, upstream changed DomainMapping to have tls field) in the meanwhile so we should re-consider about the custom TLS support again.
The custom OpenShift Route is kind of escape hatch so I want to avoid further expansion if possible.
I will create a JIRA for it.

Reply:
As you and Marcus are saying, the `disableRoute` is essentially an escape hatch, a hack of sorts. 

We should clearly state that in the docs. A clear disclaimer like this (cc @abrennan89):

> NOTE: The `serving.knative.openshift.io/disableRoute` is a simple mechanism that suppress Serverless from creating Routes, that are being created by default. Keep in mind that Serverless will keep listing that default Route URL, even if you fail to create it manually, and in that case it will not be functional. 

Also, I think your advice should be included in this page:

> If you want to configure custom domain + custom cert, we need DomainMapping as well as adding the `serving.knative.openshift.io/disableRoute: true` annotation to DomainMapping.

This is really valuable information, that's now missing. Let's write it nicely and it! WDYT @abrennan89 ?

Reply:
I think we should open a follow up Jira with your suggestions. Is that alright @cardil  ?

Reply:
Sure, we can follow up. If that's how @abrennan89 would like to do it.

Reply:
Thanks folks, I was finding this thread difficult to follow so it might make more sense to do that.